src/map/instance.cpp

```diff
@@ -257,3 +259,130 @@ bool CInstance::CheckFirstEntry(uint32 id)
     //insert returns a pair (iterator,inserted)
     return m_enteredChars.insert(id).second;
 }
+
+void CInstance::DisengageAll()
+{
+    for (EntityList_t::const_iterator it = m_charList.begin(); it != m_charList.end(); ++it)
+    {
+        CCharEntity* PCurrentChar = (CCharEntity*)it->second;
+
+        if(PCurrentChar->isAlive() && PCurrentChar->GetBattleTargetID() > 0)
+        {
+            PCurrentChar->PAI->Disengage();
+        }
+    }
+
+    for (EntityList_t::const_iterator it = m_allyList.begin(); it != m_allyList.end(); ++it)
+    {
+        CMobEntity* PCurrentAlly = (CMobEntity*)it->second;
+
+        if(PCurrentAlly->isAlive() && PCurrentAlly->GetBattleTargetID() > 0)
+        {
+            PCurrentAlly->PEnmityContainer->Clear();
+            PCurrentAlly->PAI->Disengage();
+            PCurrentAlly->PAI->Inactive(std::chrono::milliseconds(10000), false);
+        }
+    }
+
+    for (EntityList_t::const_iterator it = m_mobList.begin(); it != m_mobList.end(); ++it)
+    {
+        CMobEntity* PCurrentMob = (CMobEntity*)it->second;
+
+        if(PCurrentMob->isAlive() && PCurrentMob->GetBattleTargetID() > 0)
+        {
+            PCurrentMob->m_Aggro = false;
+            PCurrentMob->PEnmityContainer->Clear();
+            PCurrentMob->PAI->Disengage();
+            PCurrentMob->PAI->Inactive(std::chrono::milliseconds(10000), false);
+        }
+    }
+}
+
+void CInstance::StartAllyAssist(ALLY_ASSIST_MODE mode)
+{
+    // If there's no mobs, we can't assist. Don't waste time.
```

Comment:
I added both versions, and indeed it is largely copy/paste :-( 

The data structures aren't the same - instances are using std::maps everywhere, and battlefields are using std::vectors everywhere, meaning I couldn't share the code without refactoring or repeatedly converting all data sets at runtime. 

This is ultimately triggered by a generic lua function at the appropriate time in the scripts. I had bound it to this in core for now. For doing more than this directly in lua, more bindings would have to be added, which may not necessarily be a bad thing.  There is no command to directly engage a given mob in lua for example. When I looked at the Dawn battlefield with Prishe assisting, this was done by adding enmity to her on the mob. However, that doesn't work in the ToAU battlefields. Prishe is technically classified as an NPC, and these are mobs with allegiance flag = player. That trick doesn't seem to work in this case - the mobs will never engage initially no matter how much enmity you dump on them. It was also more scripting than I wanted for repeated use, so I tried to make a single trigger here to minimize what someone would have to do to get it going. 

We could definitely push more of it out to lua if we added more bindings, but I wasn't sure of where the line should be as to how much of core we expose to lua, in that once scripts start using things then refactoring the previously private behavior now has script breaking consequences.